File: clients/ruby/model/unit_action.rb
require_relative 'vec2_double'
class UnitAction
    attr_accessor :velocity
    attr_accessor :jump
    attr_accessor :jump_down
    attr_accessor :aim
    attr_accessor :shoot
    attr_accessor :swap_weapon
    attr_accessor :plant_mine
    def initialize(velocity, jump, jump_down, aim, shoot, swap_weapon, plant_mine)
        @velocity = velocity
        @jump = jump
        @jump_down = jump_down
        @aim = aim
        @shoot = shoot
        @swap_weapon = swap_weapon
        @plant_mine = plant_mine
    end
    def self.read_from(stream)
        velocity = stream.read_double()
        jump = stream.read_bool()
        jump_down = stream.read_bool()
        aim = Vec2Double.read_from(stream)
        shoot = stream.read_bool()
        swap_weapon = stream.read_bool()
        plant_mine = stream.read_bool()
        UnitAction.new(velocity, jump, jump_down, aim, shoot, swap_weapon, plant_mine)
    end
    def write_to(stream)
        stream.write_double(@velocity)
        stream.write_bool(@jump)
        stream.write_bool(@jump_down)
        @aim.write_to(stream)
        stream.write_bool(@shoot)
        stream.write_bool(@swap_weapon)
        stream.write_bool(@plant_mine)
    end
end

File: clients/go/model/unit_action.go
package model

import (
	"io"

	mStream "../stream"
)

type UnitAction struct {
	Velocity   float64
	Jump       bool
	JumpDown   bool
	Aim        *Vec2Float64
	Shoot      bool
	SwapWeapon bool
	PlantMine  bool
}

func NewUnitAction(velocity float64, jump bool, jumpDown bool, aim *Vec2Float64, shoot bool, swapWeapon bool, plantMine bool) *UnitAction {
	return &UnitAction{
		Velocity:   velocity,
		Jump:       jump,
		JumpDown:   jumpDown,
		Aim:        aim,
		Shoot:      shoot,
		SwapWeapon: swapWeapon,
		PlantMine:  plantMine,
	}
}
func ReadUnitAction(reader io.Reader) *UnitAction {
	return &UnitAction{
		Velocity:   mStream.ReadFloat64(reader),
		Jump:       mStream.ReadBool(reader),
		JumpDown:   mStream.ReadBool(reader),
		Aim:        ReadVec2Float64(reader),
		Shoot:      mStream.ReadBool(reader),
		SwapWeapon: mStream.ReadBool(reader),
		PlantMine:  mStream.ReadBool(reader),
	}
}
func (value UnitAction) Write(writer io.Writer) {
	mStream.WriteFloat64(writer, value.Velocity)
	mStream.WriteBool(writer, value.Jump)
	mStream.WriteBool(writer, value.JumpDown)
	value.Aim.Write(writer)
	mStream.WriteBool(writer, value.Shoot)
	mStream.WriteBool(writer, value.SwapWeapon)
	mStream.WriteBool(writer, value.PlantMine)
}
